### PR TITLE
TCP Socket version of the pipe demo

### PR DESCRIPTION
**Pull request by karknu**

### CODE REVIEW

File: src/Ouroboros/Network/Socket.hs [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+                        => QSem
+                        -> (BS.ByteString -> STM ())
+                        -> STM BS.ByteString
+                        -> Protocol smsg rmsg ()
+                        -> IO ()
+runProtocolWithTBQueues sem wqueue rqueue p =
+    unProtocol p >>= go mempty
+  where
+    go trailing (Send msg k) = do
+      let body = BL.toStrict $ CBOR.toLazyByteString (encode msg)
```
Comment (coot) [inline src/Ouroboros/Network/Socket.hs]: `BL.toStrict` is expensive, is it necessary?

File: src/Ouroboros/Network/Socket.hs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+instance Serialise MsgPing where
+  encode _ = encodeNull
+  decode = MsgPing <$ decodeNull
+
+
+
+runProducer :: forall block r. (Chain.HasHeader block, Serialise block)
+         => Socket -> ProducerHandlers block IO r -> IO [Async ()]
+runProducer sd producer = do
+    sem <- newQSem 0
```
Comment (coot) [inline src/Ouroboros/Network/Socket.hs]: Why do we need to use semaphores? `STM` is not enough?

Comment (karknu) [reply src/Ouroboros/Network/Socket.hs]: The semaphore is used to to implement blocking read on multiple bounded queues. It is not possible to do that using `STM` alone.

File: src/Ouroboros/Network/Socket.hs (hunk 1 above)
Comment (karknu) [reply src/Ouroboros/Network/Socket.hs]: I'm not sure that it is.

**Comment by karknu** [comment]
I'm going to change this to use `ByteChannel` instead.

**Pull request closed without merging**